### ISSUE TITLE
Parameterize on dimensions

### ISSUE THREAD

**Issue by johnmcfarlane** [g-truc/glm#584]
You [mentioned](https://github.com/g-truc/glm/issues/532#issuecomment-236038219) making C++11 the minimum spec. This would make a lot of things far easier. 

One missing feature which I've heard [mentioned](https://groups.google.com/a/isocpp.org/d/msg/sg14/68fAAq0-71k/0CCyt5o4CAAJ) and which I've been able to [implement](https://github.com/johnmcfarlane/crag/blob/develop/src/geom/Vector.h#L16) myself is a *dimension parameter*.

It would be nice to replace
```c++
template<typename T, precision P> struct tvec3;
```

with:
```c++
template<typename T, int N, precision P> struct tvec;
template<typename T, precision P> using tvec3 = tvec<T, 3, P>;
```

This would help reduce a lot of duplication, notably in the [heterogeneous operation](https://github.com/johnmcfarlane/glm/pull/1) change that I've part implemented. 

I'm not sure that this needs `using` aliases and therefore that you'd need to drop C++03. I see that you've used nested templates ([e.g.](https://github.com/g-truc/glm/blob/master/glm/detail/func_exponential.hpp#L42)) to make the code nice and generic. It may also be possible to get around the limitations of `typedef` with inheritance, e.g.
```c++
template<typename T, precision P> struct tvec3 : tvec<T, 3, P> { ... };
```

but it's hard to say without some costly experimentation. I've started experimenting in my fork with the `using` solution above and it's a big, big change!

I'm sure you've thought about this feature. Is there any particular reason not to do it? Do you think it's dependent on C++11? Would you consider a PR from me if I was able to achieve it?

**Comment by Groovounet** [comment]
Hi,

Alias templates are not supported by Visual Studio 2013 and I think it's too early to drop 2013. Actually GLM 0.9.9 is the first version that will drop Visual Studio 2010 support...

So I should have said, I am considering requiring some C++11 features.

Thanks,
Chirstophe

**Comment by johnmcfarlane** [comment]
FYI, I got this working on Clang / GCC: https://github.com/johnmcfarlane/glm/pull/2

**Comment by Groovounet** [comment]
I pulled the branch into "dim" branch to run the tests.
https://travis-ci.org/g-truc/glm/builds/187434972

With Visual C++ it doesn't built but looking at the code it looks very promising! 

I'll have a closer look.

**Comment by johnmcfarlane** [comment]
Thanks. Surprisingly, I was able to compile on VS2013 & VS2015 with very few fixes. They are pushed: https://github.com/johnmcfarlane/glm/pull/2

**Comment by Groovounet** [comment]
This issue is merged in master branch for 0.9.9 release.

Thanks for contributing,
Christophe

**Comment by johnmcfarlane** [comment]
Glad to help, thanks! (I might also take a look at `tmat` some time.)
